go-controller/pkg/node/helper_linux.go

```diff
@@ -21,44 +22,77 @@ func getDefaultGatewayInterfaceDetails() (string, []net.IP, error) {
 	var intfName string
 	var gatewayIPs []net.IP
 
-	needIPv4 := config.IPv4Mode
-	needIPv6 := config.IPv6Mode
-	routes, err := netlink.RouteList(nil, syscall.AF_UNSPEC)
-	if err != nil {
-		return "", nil, fmt.Errorf("failed to get routing table in node")
+	if config.IPv4Mode {
+		intfIPv4Name, gw, err := getDefaultGatewayInterfaceByFamily(netlink.FAMILY_V4)
+		if err != nil {
+			return "", gatewayIPs, err
+		}
+		intfName = intfIPv4Name
+		gatewayIPs = append(gatewayIPs, gw)
 	}
 
-	for _, route := range routes {
-		if route.Dst == nil && route.Gw != nil && route.LinkIndex > 0 {
-			intfLink, err := netlink.LinkByIndex(route.LinkIndex)
+	if config.IPv6Mode {
+		intfIPv6Name, gw, err := getDefaultGatewayInterfaceByFamily(netlink.FAMILY_V6)
+		if err != nil {
+			return "", gatewayIPs, err
+		}
+		// validate that both IP Families use the same interface for the gateway
+		if intfName == "" {
+			intfName = intfIPv6Name
+		} else if intfName != intfIPv6Name {
+			return "", nil, fmt.Errorf("multiple gateway interfaces detected: %s %s", intfName, intfIPv6Name)
+		}
+		gatewayIPs = append(gatewayIPs, gw)
+	}
+
+	if len(gatewayIPs) == 0 {
+		return "", nil, fmt.Errorf("failed to get default gateway interface")
+	}
+	return intfName, gatewayIPs, nil
```

Comment:
Will it be the case that len(gatewayIPs) == 0 ? If we don't get a default gateway interface either for IPv4 or IPv6, then we return with failure before we hit L48, isn't it? In other words, can we remove L48-50?

Reply:
we need to return an error, don´t we?

Reply:
yep, we error so the pod restarts. I think that we have to keep it this way

Reply:
Right.  However, my comment was that at L48 it will never be the case that `len(gatewayIPs) = 0` ever and that check is redundant.

Reply:
if config.IPv4Mode == config.IPv6Mode == false it will be 0
I always try to be pedantic and not assume that the functions that we call will do the right thing ;), in this case `getDefaultGatewayInterfaceByFamily()` , I wanted to add some unit tests but the netlink dependency complicated it a bit :/